# Architecture of androgen receptor pathways amplifying glucagon-like peptide-1 insulinotropic action in male pancreatic β-cells

# Table of Contents
- [What is this?](#what-is-this)
- [How can I use this data, and where can I find it?](#how-can-i-use-this-data-and-where-can-i-find-it)
	- [Downloading Data files](#downloading-data-files)
	- [Data sub-structure](#data-sub-structure)
		- [FASTQ files](#fastq-files)
		- [Cellranger output (processed gene-counts of single cells/barcodes)](#cellranger-output-processed-gene-counts-of-single-cellsbarcodes)
- [Analysis and visualization programs](#analysis-and-visualization-programs)
	- [Cellranger software from 10X Genomics:](#cellranger-software-from-10x-genomics)
	- [R and R's integrated developmental environment RStudio:](#r-and-rs-integrated-developmental-environment-rstudio)
	- [scRNAseq analysis pipeline SEURAT developed by the Satija lab:](#scrnaseq-analysis-pipeline-seurat-developed-by-the-satija-lab)
	
- [Setting up the right environment](#setting-up-the-right-environment)
- [Computing environment](#computing-environment)
- [Contributors](#contributors)
- [Lead Contacts](#lead-contacts)


## What is this?
This repository contains coding scripts utilized for the analysis performed in the "Architecture of androgen receptor pathways amplifying glucagon-like peptide-1 insulinotropic action in male pancreatic β-cellse" publication (Xu-Qadir et. al, 2023). The purpose of providing the code here is to allow for transparency and robust data-analysis reproducibility. Most of the steps used for data analysis and visualization have been optimised for an average computing environment (for the year 2023). Some analyses however, require a high-performace computing environment (see computing environment). The methodology has already been described extensively in the manuscript. However, this analysis relies heavily on powerful scRNAseq analysis algorithms developed by the [Satija lab](https://satijalab.org/), namely [Seurat](https://satijalab.org/seurat/) [(Butler et al., 2018: Nature Biotechnology;](https://www.nature.com/articles/nbt.4096) [Stuart et al., 2018: Cell)](https://www.sciencedirect.com/science/article/pii/S0092867419305598?via%3Dihub) (for a complete list of dependencies and code utilized see analysis & visualization programs).

## How can I use this data, and where can I find it?
### Downloading Data files
Data files utilized in this analysis have been deposited in the [Gene Expression Omnibus (GEO)](https://www.ncbi.nlm.nih.gov/geo/), gene expression data repository at the NIH. Data are part of the GSE131886 high-thoroughput sequencing repository and can be found [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE201256). Contact lead author for seurat object.

### Data sub-structure
We povide raw FASTQ files generated from single-cell cDNA libraries sequenced by the Illumina sequencing platform, along with unfiltered post-alignment count files generated by the Cellranger software. In addition we also provide a gene expression matrix containing data on filtered gene counts across our dataset.

#### FASTQ files
These are sequencing reads generated by the Illumina sequencing platform. Files contain raw reads and sequencing efficiency information.
These are the input files for the Cellranger software. and can be found here: [GSE201256](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE201256)

#### Cellranger output (processed gene-counts of single cells/barcodes)
This contains data outputs of Cellranger, which was run using default settings. Code used to analyze data is a part of this repository. This data contains filtered/unfiltered count files for gene expression across barcodes/cells. 

#### Data analysis of this dataset
Preliminary data-analyses involving n=3 ctrl + n=3 DHT Tx deidentified human cadaveric Islet samples.

## Analysis and visualization programs
### Cellranger software from 10X Genomics:
1. [Cellranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/installation)

### R and R's integrated developmental environment RStudio:
1. [R v4.2.2 (x64 bit)](https://cran.r-project.org/bin/windows/base/old/)
2. [RStudio build #386 (x64 bit)](https://www.rstudio.com/products/rstudio/download/)
3. [RTools v4.2.X](https://cran.r-project.org/bin/windows/Rtools/index.html)
4. [Tutorial for R](https://cran.r-project.org/doc/manuals/r-release/R-intro.html)
5. [Tutorial for RStudio](https://resources.rstudio.com/) 

### scRNAseq analysis pipeline SEURAT developed by the Satija lab:
1. [Source code for Seurat](https://cran.r-project.org/web/packages/Seurat/index.html)
2. [Tutorials for Seurat](https://satijalab.org/seurat/)


## Setting up the right environment
1. Install R
2. Install Rstudio
3. Once you have installed R and RStudio, you can run the script.
4. If you need help understanding how commands are run in R  use the [ctrl + enter] command or please visit [here](https://support.rstudio.com/hc/en-us/articles/200484448-Editing-and-Executing-Code).
6. If you run into problems, please open a new issue, you can do this by going to 'issues' and clicking on the 'new issue' icon.

## Computing environment
### Hardware and OS environment for running analysis
#### Environment
1. Processor: Intel core i9-9900 (16cores x 16 threads)
2. RAM: 128GB DDR3
3. OS: Windows 10 Enterprise (x64 bit)


# Citation
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.7814544.svg)](https://doi.org/10.5281/zenodo.7814544)
-
Xu, W., Qadir, F., Nasteska, D., Mota de Sa, P., Gorvin, C. M., Blandino-Rosano, M., ... & Mauvais-Jarvis, F. Molecular Architecture of Androgen Receptor Pathways Amplifying Glucagon-Like Peptide-1 Insulinotropic Action in Male Pancreatic β-Cells. [Cell Reports](https://www.cell.com/cell-reports/fulltext/S2211-1247(23)00540-5).

# Contributors
1. Mirza Muhammad Fahd Qadir - [Github](https://github.com/Dragonmasterx87) - to contact please [Email](mailto:fahd.qadir@med.miami.edu)

# Lead Contacts
1. Dr. Franck Mauvais-Jarvis MD PhD. - [Dept of Medicine, Tulane University](https://medicine.tulane.edu/departments/medicine-endocrinology-metabolism-diabetes-research-tulane-center-aging-tips-mentor) - to contact please [Email](mailto:fmauvais@tulane.edu)

# Acknowledgements
1. NIDDK
2. VA

## If this was useful to you please dont forget to cite, star and fork this repository.
